flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java

```diff
@@ -0,0 +1,438 @@
+                        .thenApply(CompletedCheckpoint::getExternalPointer);
+
+        final CompletableFuture<JobStatus> terminationFuture =
+                executionGraph
+                        .getTerminationFuture()
+                        .handle(
+                                (jobstatus, throwable) -> {
+                                    if (throwable != null) {
+                                        logger.info(
+                                                "Failed during stopping job {} with a savepoint. Reason: {}",
+                                                executionGraph.getJobID(),
+                                                throwable.getMessage());
+                                        throw new CompletionException(throwable);
+                                    } else if (jobstatus != JobStatus.FINISHED) {
+                                        logger.info(
+                                                "Failed during stopping job {} with a savepoint. Reason: Reached state {} instead of FINISHED.",
+                                                executionGraph.getJobID(),
+                                                jobstatus);
+                                        throw new CompletionException(
+                                                new FlinkException(
+                                                        "Reached state "
+                                                                + jobstatus
+                                                                + " instead of FINISHED."));
+                                    }
+                                    return jobstatus;
+                                });
+
+        return savepointFuture
+                .thenCompose((path) -> terminationFuture.thenApply((jobStatus -> path)))
+                .handleAsync(
+                        (path, throwable) -> {
+                            if (throwable != null) {
+                                if (context.isState(this)) {
+                                    // restart the checkpoint coordinator if stopWithSavepoint
+                                    // failed.
+                                    startCheckpointScheduler(checkpointCoordinator);
+                                }
+                                throw new CompletionException(throwable);
+                            }
+                            if (ensureAllExecutionsFinished()) {
+                                Throwable error =
+                                        new FlinkException(
+                                                "Not all executions were stopped with a savepoint.");
+                                handleGlobalFailure(error);
+                                throw new CompletionException(error);
+                            }
+
+                            return path;
+                        },
+                        context.getMainThreadExecutor());
+    }
```

Comment:
I would suggest to solve this problem properly with an explicit `State` which marks that we are stopping with a savepoint. Hence, we could implement this method with an `UnsupportedOperationException` and then do this as a follow up.

Reply:
Okay. I filed a ticket: https://issues.apache.org/jira/browse/FLINK-21333